src/background/adblocker.js

```diff
@@ -379,56 +221,119 @@ async function executeScriptlets(tabId, frameId, scripts) {
+
+  chrome.scripting
+    .insertCSS({
+      css: styles,
+      origin: 'USER',
+      target,
+    })
+    .catch((e) => console.warn('[adblocker] failed to inject CSS', e));
+}
+
+async function injectCosmetics(details, config) {
   try {
     setup.pending && (await setup.pending);
   } catch (e) {
-    console.error('[adblocker] not ready for scriptlet injection', e);
+    console.error('[adblocker] not ready for cosmetic injection', e);
     return;
   }
 
-  const { hostname, domain } = parse(url);
-  if (!hostname || isPaused(options, hostname)) {
-    return;
-  }
+  const { frameId, url, tabId } = details;
+
+  const parsed = parse(url);
+  const domain = parsed.domain || '';
+  const hostname = parsed.hostname || '';
+
+  if (isPaused(options, hostname)) return;
 
   const tabHostname = tabStats.get(tabId)?.hostname;
   if (tabHostname && isPaused(options, tabHostname)) {
     return;
   }
 
   const engine = engines.get(engines.MAIN_ENGINE);
+  const isBootstrap = config.bootstrap;
 
-  const { scripts } = engine.getCosmeticsFilters({
-    url: url,
-    hostname,
-    domain: domain || '',
-    getBaseRules: false,
-    getInjectionRules: true,
-    getExtendedRules: false,
-    getRulesFromDOM: false,
-    getRulesFromHostname: true,
-  });
-
-  if (scripts.length > 0) {
-    executeScriptlets(tabId, frameId, scripts);
-  }
-}
+  {
+    const cosmetics = engine.getCosmeticsFilters({
+      domain,
+      hostname,
+      url,
+
+      classes: config.classes,
+      hrefs: config.hrefs,
+      ids: config.ids,
 
-if (__PLATFORM__ === 'safari') {
-  chrome.runtime.onMessage.addListener((msg, sender) => {
-    if (sender.url && msg.action === 'injectScriptlets') {
-      injectScriptlets(sender.tab.id, sender.frameId, sender.url);
+      getBaseRules: false,
+      // This needs to be done only once per frame
+      getInjectionRules: isBootstrap,
+      getExtendedRules: isBootstrap,
+      getRulesFromHostname: isBootstrap,
+
+      // This will be done every time we get information about DOM mutation
+      getRulesFromDOM: !isBootstrap,
+    });
+
+    if (isBootstrap && cosmetics.scripts.length > 0) {
+      injectScriptlets(cosmetics.scripts, tabId, frameId);
     }
 
-    return false;
-  });
+    if (cosmetics.styles) {
+      injectStyles(cosmetics.styles, tabId, frameId);
+    }
+  }
+
+  if (frameId === 0 && isBootstrap) {
+    const { styles } = engine.getCosmeticsFilters({
+      domain,
+      hostname,
+      url,
+
+      // This needs to be done only once per tab
+      getBaseRules: true,
+      getInjectionRules: false,
+      getExtendedRules: false,
+      getRulesFromDOM: false,
+      getRulesFromHostname: false,
+    });
+
+    injectStyles(styles, tabId);
+  }
 }
 
-chrome.webNavigation.onCommitted.addListener((details) => {
-  injectScriptlets(details.tabId, details.frameId, details.url);
+chrome.webNavigation.onCommitted.addListener(
+  (details) => {
+    injectCosmetics(details, { bootstrap: true });
```

Comment:
This is completely optional suggestion but is there any reason to wrap this with an object? Can we just pass `true` instead of `{ bootstrap: true }`? I don't think we need to care about extensibility here.

Reply:
It takes more parameters, look at the https://github.com/ghostery/ghostery-extension/blob/3b70ef64c69ac77d23e6870472482ef06e9ffa7b/src/background/adblocker.js#L272

Reply:
Oh, I didn't notice those are coming from `config`. Then all good, thanks for the explanation 👍 .